### PR TITLE
fixes #4: implement extension API

### LINKED ISSUE #4: implement extension API
Example taken from the SuperPack specification:

``` js
SuperPackTranscoder.extend(
  // extension point: 0 through 127
  0,
  // detect values which require this custom serialisation
  x => x instanceof RegExp,
  // serialiser: return an intermediate value which will be encoded instead
  r => [r.pattern, r.flags],
  // deserialiser: from the intermediate value, reconstruct the original value
  ([pattern, flags]) => RegExp(pattern, flags),
);
```

We will probably want an OO-ish interface to allow this.

### PR DESCRIPTION
**Pull request by michaelficarra**
Fixes #4. No extension tests yet, so please do not merge. Required a huge refactoring. I want to make sure we settle https://github.com/shapesecurity/superpack-spec/issues/7 before merging this as well.

### CODE REVIEW

**Comment by michaelficarra** [comment]
@disnet Ready for review.

**Comment by michaelficarra** [comment]
@disnet I can't repro the Travis failure locally. It always passes for me. Can you clone the repo and run the tests on your machine? It's behaving as if the `flags` field of a regexp instance is an undefined value. But it's certainly not. It's always a string.

**Comment by disnet** [comment]
I can confirm it works locally on my machine but my machine is not running IE or edge. Those are the browsers failing.

**Comment by michaelficarra** [comment]
CI passed! Please review @disnet.

File: src/decoder.js [hunk 1]
```diff
@@ -1,221 +1,246 @@
...
-  } else {
-    stringLUT = [];
-    keysetLUT = [];
+  static decode(buffer : SuperPackedValue, options? : { omittedKeysets? : Array<Keyset>, extensions? : ExtensionMap } = {}) : any {
+    let d = new Decoder;
+    if (options.extensions != null) {
+      // $FlowFixMe: flow doesn't understand that ext is an ExtensionPoint
+      Object.keys(options.extensions).forEach((ext : ExtensionPoint) => {
+        // $FlowFixMe: flow doesn't understand that options.extensions is non-null here
+        let extension = options.extensions[ext];
```
Comment (disnet) [inline src/decoder.js]: I think you can solve this by doing `const extensions = options.extensions` before the `if`. Flow doesn't know that `options` hasn't been mutated.

**Pull request merged**